File: packages/api/go/core/objects/id.go
package objects

import (
	"strings"

	"github.com/google/uuid"

	"template/core-go/errors"
)

// idNamespace is the UUIDv5 namespace for every canonical entity ingested
// from external systems. Locked byte-for-byte per spec §"Deterministic IDs".
// Mirrored on the TS side as the BK_DASH_NAMESPACE constant in
// packages/api/typescript/core/src/objects/Id.ts — any change here MUST be
// reflected there and vice versa, otherwise previously-ingested entities
// orphan across the polyglot boundary. Not exported — callers go through
// IDFromSeed.
var idNamespace = uuid.MustParse("f63cfbe6-7d8e-4d1d-9c5d-9d8a8b1b0c5e")

// ID is a value object wrapping a UUID. It supports:
// - Random generation: NewID()
// - Pre-set value: IDFromString(s), IDFromUUID(u)
// - Deterministic from seed parts: IDFromSeed(parts...) — UUIDv5 from input strings
type ID struct {
	value uuid.UUID
}

// NewID generates a new random UUID.

func NewID() ID {
	return ID{value: uuid.New()}
}

// IDFromUUID wraps an existing uuid.UUID.

func IDFromUUID(u uuid.UUID) ID {
	return ID{value: u}
}

// IDFromString parses a string into an ID.

func IDFromString(s string) (ID, error) {
	u, err := uuid.Parse(s)
	if err != nil {
		return ID{}, errors.NewBaseError(errors.CodeInvalidID, "invalid ID format: "+s)
	}
	return ID{value: u}, nil
}

// IDFromSeed creates a deterministic UUIDv5 from the provided parts, scoped
// to idNamespace. Parts are joined by `:` before hashing — same
// separator + namespace as the TS side (`Id.fromSeed(...)`), so identical
// inputs produce identical UUIDs across languages.
//
// Example: IDFromSeed("resource", "external-key") produces the same UUID on
// both TS + Go for identical inputs.
func IDFromSeed(parts ...string) (ID, error) {
	if len(parts) == 0 {
		return ID{}, errors.NewBaseError(errors.CodeInvalidID, "at least one part is required for IDFromSeed")
	}

	u := uuid.NewSHA1(idNamespace, []byte(strings.Join(parts, ":")))
	return ID{value: u}, nil
}


func (id ID) UUID() uuid.UUID { return id.value }


func (id ID) Value() string { return id.value.String() }

func (id ID) String() string       { return id.value.String() }

func (id ID) Equals(other ID) bool { return id.value == other.value }

func (id ID) IsZero() bool         { return id.value == uuid.Nil }

// Bytes returns the raw 16 bytes of the underlying UUID.

func (id ID) Bytes() [16]byte { return id.value }

File: packages/api/typescript/core/src/objects/Id.ts
import { BasePrimitiveValueObject } from './BasePrimitiveValueObject'
import type { BaseDomainErrors } from '../errors/codes'
import { createHash } from 'node:crypto'
import { uuidv7 } from 'uuidv7'
import z from 'zod'

export const IdSchema = z.string().refine(v => v.length > 0, { error: 'INVALID_ID' as BaseDomainErrors })

// Locked UUIDv5 namespace — byte-for-byte identical to `packages/api/go/core/objects/id.go`.
// Changing this orphans every previously-ingested deterministic id. Not exported.
const ID_NAMESPACE = 'f63cfbe6-7d8e-4d1d-9c5d-9d8a8b1b0c5e' as const

export class Id extends BasePrimitiveValueObject<typeof IdSchema> {
	static override schema = IdSchema

	constructor(value?: string) {
		super(value ?? Id.value())
	}

	/**
	 * Fresh primitive id string (UUIDv7). Use when an external library
	 * needs a raw id generator and we don't want to leak uuidv7 imports
	 * across the codebase.
	 */
	static value(): string {
		return uuidv7()
	}

	/**
	 * Deterministic UUIDv5 from a tuple of strings, using the locked
	 * `ID_NAMESPACE`. Same algorithm + namespace as Go's
	 * `objects.IDFromSeed(...)` — identical inputs produce identical UUIDs
	 * across languages.
	 *
	 * Implementation: RFC 4122 UUIDv5 = SHA-1(namespace_bytes || name_bytes)
	 * with version + variant bits set per spec. Inlined to avoid a `uuid`
	 * package dependency.
	 */
	static fromSeed(...parts: string[]): Id {
		if (parts.length === 0) throw new Error('Id.fromSeed requires at least one part')
		const name = parts.join(':')

		const namespaceBytes = Buffer.from(ID_NAMESPACE.replace(/-/g, ''), 'hex')
		const hash = createHash('sha1').update(namespaceBytes).update(name, 'utf8').digest()

		// Set UUIDv5 version (0x50 in byte 6) and RFC 4122 variant (0x80 in byte 8).
		hash[6] = (hash[6]! & 0x0f) | 0x50
		hash[8] = (hash[8]! & 0x3f) | 0x80

		const hex = hash.subarray(0, 16).toString('hex')
		const uuid = `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-` + `${hex.slice(16, 20)}-${hex.slice(20, 32)}`
		return new Id(uuid)
	}

	equals(other: Id): boolean {
		return this.value === other.value
	}

	override toString(): string {
		return this.value
	}
}
